rfc/RFC-ABM-210329.md

```diff
@@ -0,0 +1,143 @@
+        Source                    Video Player     IPFS Client
+          |                              ^             ^                   ^             ^
+    (rtmp-publish)                       |             |                   |             |
+          |                            (http)        (ipfs)              (http)        (ipfs)
+          v                              |             |                   |             |
++-------------------+                +---------------------+           +---------------------+
+| rtmp-server       |                | ipfs-server         |           | ipfs-mirror         |
+| ↳ nginx-rtmp      |                | ↳ ipfs with pubsub  |           | ↳ ipfs with pubsub  |
+|                   |<--(rtmp-pull)--| ↳ ipfs-http gateway |<--(ipfs)--| ↳ ipfs-http gateway |
+|                   |                | ↳ ffmpeg            |           |- - - - - - - - - - -|
+|- - - - - - - - - -|                |- - - - - - - - - - -|           | Pins IPFS hashes    |
+| Runs RTMP server  |                | Encodes HLS ts+m3u8 |           | learnt from IPNS id |
+| publishable from  |                | pins on IPFS and    |           | of ipfs-server      |
+| authenticated IPs |                | publishes to IPNS   |           +---------------------+
++-------------------+                +---------------------+
+          |
+   (rtmp-pull/push)
+          |
+          v
+Other Streaming Services
+```
+
+The on-premise network camera pushes to the `rtmp-server`.
+The `ipfs-server` pulls that RTMP stream, encodes ts chunks in a live m3u8 file using
+ffmpeg, then IPFS adds and pins those files and uses IPNS to publish the m3u8 to its node
+ID. The built-in ipfs-http gateway allow those content to be accessed via HTTP,
+which is what the embedded player on the website will use.
+However, viewers running an IPFS client (with pubsub enabled) can directly view the streams over IPFS.
+Optionally, we can run one or more `ipfs-mirror` servers that pin the live streaming content and run additional gateways.
+
+
+### RTMP server
+Real-Time Messaging Protocol (RTMP) is an open source protocol owned by Adobe that's designed to stream audio and video by maintaining low latency connections. Clients use a handshake to form a connection with an RTMP server which then allows users to stream video and audio.
+
+
+## Implementation details
+
+- Implementation languages: `Go` and `javascript`.
+
+- Server prerequisites:
+	- install `ffmpeg`
+	- install `go-ipfs` and start `IPFS daemon`
+
+- Server-side implementation layout
+	- FFMpeg controller: 
+		- connects the network camera and starts encoding to publishable format such as the `ts+m3u8` format combination that's suitable for HLS players embedded in mainstream browsers.
+	- File watcher:
+		- whenever there's a new stream file generated the watcher notifies the observers to update information accordingly
+	- IPFS controller:
+		- adds and pins data on the network
```

Comment:
I problem that why we need IPNS, if user want watch history video, he must to known which file to load, so he will query the file from server, then get file from ipfs

Reply:
IPNS publishes mutable content to the network with an immutable hash; users on the client-side can then resolve the immutable hash to get the newest content.